## 첫구매 주문 여부와, 첫 구매 후 재구매까지 걸린 시간 

In [150]:
import pandas as pd
import numpy as np
%matplotlib inline
import matplotlib.pyplot as plt
plt.rcParams["font.family"] = 'Malgun Gothic'
plt.rcParams.update({'font.size': 25})

In [131]:
sales=pd.read_csv("../../data/sales/3.sample data.csv",encoding='ms949')

## EDA

In [132]:
# 주문 고객 수
sales['아이디'].drop_duplicates().shape

(17805,)

In [133]:
# 주문 수
sales.shape

(43189, 3)

In [134]:
# 주문 기간 
sales.sort_values(['결제일'])

주문번호    아이디              결제일
7301    7302  maa00  2020-01-01 0:01
7300    7301  yha50  2020-01-01 0:07
7299    7300  vna12  2020-01-01 0:11
7297    7298  kaa8c  2020-01-01 0:28
7296    7297  cea35  2020-01-01 0:37
...      ...    ...              ...
36245  36246  ena03  2020-05-31 9:48
36244  36245  kkaaz  2020-05-31 9:49
36243  36244  ematy  2020-05-31 9:54
36242  36243  nha75  2020-05-31 9:57
36241  36242  nha67  2020-05-31 9:58

[43189 rows x 3 columns]

## 1. 첫 주문인지 확인 하기

In [135]:
sales['결제일']=pd.to_datetime(sales['결제일'],infer_datetime_format=True)
day_group=sales.groupby('아이디')['결제일']

In [136]:
sales['첫 주문일']=day_group.transform(np.min)
sales.loc[sales['결제일'] == sales['첫 주문일'],'첫주문여부'] = True
sales.loc[sales['결제일'] != sales['첫 주문일'],'첫주문여부'] = False

In [137]:
sales.head()

주문번호    아이디                 결제일               첫 주문일  첫주문여부
0     1  nha43 2020-01-31 23:56:00 2020-01-22 13:47:00  False
1     2  mia80 2020-01-31 23:50:00 2020-01-28 13:49:00  False
2     3  dna19 2020-01-31 23:44:00 2020-01-31 23:44:00   True
3     4  tba99 2020-01-31 23:30:00 2020-01-31 23:30:00   True
4     5  nea20 2020-01-31 23:28:00 2020-01-31 23:28:00   True

### 1) 특정 고객의 첫 주문일과 재구매 확인

- 해당 고객은 1월 22일 첫 구매를 했으며, 24번의 재구매를 했음

In [138]:
sales[sales['아이디'] =='nha43']

주문번호    아이디                 결제일               첫 주문일  첫주문여부
0          1  nha43 2020-01-31 23:56:00 2020-01-22 13:47:00  False
1138    1139  nha43 2020-01-26 14:52:00 2020-01-22 13:47:00  False
1773    1774  nha43 2020-01-22 13:47:00 2020-01-22 13:47:00   True
12958  12959  nha43 2020-02-07 14:56:00 2020-01-22 13:47:00  False
15895  15896  nha43 2020-03-25 14:34:00 2020-01-22 13:47:00  False
16410  16411  nha43 2020-03-23 16:38:00 2020-01-22 13:47:00  False
18720  18721  nha43 2020-03-17 11:13:00 2020-01-22 13:47:00  False
21962  21963  nha43 2020-04-29 13:29:00 2020-01-22 13:47:00  False
22043  22044  nha43 2020-04-29 09:59:00 2020-01-22 13:47:00  False
22154  22155  nha43 2020-04-28 21:57:00 2020-01-22 13:47:00  False
24373  24374  nha43 2020-04-25 08:31:00 2020-01-22 13:47:00  False
25152  25153  nha43 2020-04-24 13:27:00 2020-01-22 13:47:00  False
26415  26416  nha43 2020-04-21 23:05:00 2020-01-22 13:47:00  False
27368  27369  nha43 2020-04-20 02:03:00 2020-01-22 13:47:00  False
30695  30696  nha43 2020-04-03 15:25:00 2020-01-22 13:47:00  False
30905  30906  nha43 2020-04-02 10:43:00 2020-01-22 13:47:00  False
31080  31081  nha43 2020-04-01 11:58:00 2020-01-22 13:47:00  False
31411  31412  nha43 2020-05-16 01:52:00 2020-01-22 13:47:00  False
31811  31812  nha43 2020-05-14 21:59:00 2020-01-22 13:47:00  False
33178  33179  nha43 2020-05-11 13:06:00 2020-01-22 13:47:00  False
33486  33487  nha43 2020-05-09 23:56:00 2020-01-22 13:47:00  False
35395  35396  nha43 2020-05-03 02:09:00 2020-01-22 13:47:00  False
35769  35770  nha43 2020-05-01 09:57:00 2020-01-22 13:47:00  False
36359  36360  nha43 2020-05-30 22:46:00 2020-01-22 13:47:00  False
42719  42720  nha43 2020-05-18 09:35:00 2020-01-22 13:47:00  False

### 2) 월 별 첫 주문 고객 수

In [139]:
sales['결제월']=sales['결제일'].dt.month

In [140]:
sales_table=sales.pivot_table(index='결제월',columns='첫주문여부',values='아이디',aggfunc='count',margins=True)
sales_table

첫주문여부  False   True    All
결제월                       
1       1964   5355   7319
2       3578   3437   7015
3       4533   2710   7243
4       6499   3092   9591
5       8781   3240  12021
All    25355  17834  43189

첫 주문 고객의 비율

In [141]:
table2 = sales_table.div( sales_table.iloc[:,-1], axis=0 ).round(2)
table2

첫주문여부  False  True  All
결제월                    
1       0.27  0.73  1.0
2       0.51  0.49  1.0
3       0.63  0.37  1.0
4       0.68  0.32  1.0
5       0.73  0.27  1.0
All     0.59  0.41  1.0

- 1월에는 첫구매자 비율이 73%, 5355명으로 가장 많았고, 점차 재구매자가 증가해서 5월에는 재구매자 비율이 73%로 역전
- 매달 약 3,000명의 첫 구매 회원이 유입되고 있으며, 재구매자가 증가하여 전체적인 주문 볼륨이 늘어나고 있음

# 2. 재구매까지 소요되는 일자 보기

In [142]:
s_reorder=sales[(sales['첫주문여부']==False)]

In [143]:
s_reorder.head()

주문번호    아이디                 결제일               첫 주문일  첫주문여부  결제월
0      1  nha43 2020-01-31 23:56:00 2020-01-22 13:47:00  False    1
1      2  mia80 2020-01-31 23:50:00 2020-01-28 13:49:00  False    1
5      6  nhaf0 2020-01-31 23:19:00 2020-01-02 00:20:00  False    1
6      7  nhae7 2020-01-31 23:16:00 2020-01-07 19:52:00  False    1
10    11  joa72 2020-01-31 22:45:00 2020-01-09 15:11:00  False    1

In [144]:
day_group2=s_reorder.groupby('아이디')['결제일']
s_reorder['두번째 주문일']=day_group2.transform(np.min)
s_reorder.loc[s_reorder['결제일'] == s_reorder['두번째 주문일'],'두번째 주문여부'] = True
s_reorder.loc[s_reorder['결제일'] != s_reorder['두번째 주문일'],'두번째 주문여부'] = False

C:\Users\user\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [145]:
s_reorder=s_reorder[['주문번호','두번째 주문일','두번째 주문여부']]

In [146]:
sales=sales.merge(s_reorder,on='주문번호',how='left')
sales

주문번호    아이디                 결제일               첫 주문일  첫주문여부  결제월  \
0          1  nha43 2020-01-31 23:56:00 2020-01-22 13:47:00  False    1   
1          2  mia80 2020-01-31 23:50:00 2020-01-28 13:49:00  False    1   
2          3  dna19 2020-01-31 23:44:00 2020-01-31 23:44:00   True    1   
3          4  tba99 2020-01-31 23:30:00 2020-01-31 23:30:00   True    1   
4          5  nea20 2020-01-31 23:28:00 2020-01-31 23:28:00   True    1   
...      ...    ...                 ...                 ...    ...  ...   
43184  43185  kaa9d 2020-05-20 11:43:00 2020-01-21 11:34:00  False    5   
43185  43186  nhae7 2020-05-17 11:16:00 2020-01-07 19:52:00  False    5   
43186  43187  nha9a 2020-05-17 09:19:00 2020-01-01 01:04:00  False    5   
43187  43188  doa10 2020-05-20 10:25:00 2020-01-01 02:45:00  False    5   
43188  43189  kaa38 2020-05-17 01:04:00 2020-01-07 17:34:00  False    5   

                  두번째 주문일 두번째 주문여부  
0     2020-01-26 14:52:00    False  
1     2020-01-31 23:50:00     True  
2                     NaT      NaN  
3                     NaT      NaN  
4                     NaT      NaN  
...                   ...      ...  
43184 2020-01-21 16:08:00    False  
43185 2020-01-08 06:09:00    False  
43186 2020-01-01 10:18:00    False  
43187 2020-01-21 20:08:00    False  
43188 2020-01-19 20:35:00    False  

[43189 rows x 8 columns]

[정리]

1) 두번째 주문일이 있으면서, 두번째 주문여부가 True 인 주문이 2번째 구매

2) 두번째 주문일이 있으면서, 두번째 주문여부가 False 인 주문은 2번째 이후 재구매

3) 두번째 주문일이 없는 주문일 경우, 첫구매 이후 재구매가 이루어지지 않은 고객 

In [149]:
sales['재구매 소요기간'] = sales['두번째 주문일'] - sales['첫 주문일']
sales

주문번호    아이디                 결제일               첫 주문일  첫주문여부  결제월  \
0          1  nha43 2020-01-31 23:56:00 2020-01-22 13:47:00  False    1   
1          2  mia80 2020-01-31 23:50:00 2020-01-28 13:49:00  False    1   
2          3  dna19 2020-01-31 23:44:00 2020-01-31 23:44:00   True    1   
3          4  tba99 2020-01-31 23:30:00 2020-01-31 23:30:00   True    1   
4          5  nea20 2020-01-31 23:28:00 2020-01-31 23:28:00   True    1   
...      ...    ...                 ...                 ...    ...  ...   
43184  43185  kaa9d 2020-05-20 11:43:00 2020-01-21 11:34:00  False    5   
43185  43186  nhae7 2020-05-17 11:16:00 2020-01-07 19:52:00  False    5   
43186  43187  nha9a 2020-05-17 09:19:00 2020-01-01 01:04:00  False    5   
43187  43188  doa10 2020-05-20 10:25:00 2020-01-01 02:45:00  False    5   
43188  43189  kaa38 2020-05-17 01:04:00 2020-01-07 17:34:00  False    5   

                  두번째 주문일 두번째 주문여부         재구매 소요기간  
0     2020-01-26 14:52:00    False  4 days 01:05:00  
1     2020-01-31 23:50:00     True  3 days 10:01:00  
2                     NaT      NaN              NaT  
3                     NaT      NaN              NaT  
4                     NaT      NaN              NaT  
...                   ...      ...              ...  
43184 2020-01-21 16:08:00    False  0 days 04:34:00  
43185 2020-01-08 06:09:00    False  0 days 10:17:00  
43186 2020-01-01 10:18:00    False  0 days 09:14:00  
43187 2020-01-21 20:08:00    False 20 days 17:23:00  
43188 2020-01-19 20:35:00    False 12 days 03:01:00  

[43189 rows x 9 columns]

In [165]:
sales['재구매 소요시간']=(sales['재구매 소요기간'].dt.total_seconds() / 3600).round(0)

In [175]:
sales['재구매 소요시간'].describe()

count    25355.000000
mean       529.300138
std        640.362516
min          0.000000
25%         90.000000
50%        272.000000
75%        711.000000
max       3600.000000
Name: 재구매 소요시간, dtype: float64